core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BpmBillingProjectCreator.scala

```diff
@@ -0,0 +1,57 @@
+package org.broadinstitute.dsde.rawls.billing
+
+import akka.http.scaladsl.model.StatusCodes
+import org.broadinstitute.dsde.rawls.RawlsExceptionWithErrorReport
+import org.broadinstitute.dsde.rawls.model.{CreateRawlsV2BillingProjectFullRequest, ErrorReport, UserInfo}
+
+import scala.concurrent.{ExecutionContext, Future, blocking}
+
+
+/**
+ * This class knows how to validate Rawls billing project requests and instantiate linked billing profiles in the
+ * billing profile manager service.
+ */
+class BpmBillingProjectCreator(billingRepository: BillingRepository,
+                               billingProfileManagerDAO: BillingProfileManagerDAO)
+                              (implicit val executionContext: ExecutionContext) extends BillingProjectCreator {
+
+  override def validateBillingProjectCreationRequest(createProjectRequest: CreateRawlsV2BillingProjectFullRequest,
+                                                     userInfo: UserInfo): Future[Unit] = {
+    val azureManagedAppCoordinates = createProjectRequest.billingInfo match {
+      case Left(_) => throw new NotImplementedError("Google billing accounts not supported in billing profiles")
+      case Right(coords) => coords
+    }
+
+    for {
+      apps <- blocking {
+        billingProfileManagerDAO.listManagedApps(azureManagedAppCoordinates.subscriptionId, userInfo)
+      }
+      _ = apps.find(app => app.getSubscriptionId == azureManagedAppCoordinates.subscriptionId &&
+        app.getManagedResourceGroupId == azureManagedAppCoordinates.managedResourceGroupId &&
+        app.getTenantId == azureManagedAppCoordinates.tenantId
+      ) match {
+        case None => throw new ManagedAppNotFoundException(ErrorReport(StatusCodes.Forbidden, "Managed application not found"))
+        case Some(_) => ()
+      }
+    } yield {}
+  }
+
+  /**
+   * Creates a billing profile with the given billing creation info and links the previously created billing project
+   * with it
+   */
+  override def postCreationSteps(createProjectRequest: CreateRawlsV2BillingProjectFullRequest, userInfo: UserInfo): Future[Unit] = {
+    for {
```

Comment:
For/yield question again.

Reply:
So this is syntactic sugar over flatmapping. That is, we need to express a dependency between the two futures:
1) The BPM operation to create the profile
2) The repo operation to update the billing profile ID which needs the UUID returned from step 1 to establish the linkage. 

Desguared, we could do something like:

```
billingProfileManagerDAO.createBillingProfile(createProjectRequest.projectName.value, createProjectRequest.billingInfo, userInfo).flatMap {
      profileModel => billingRepository.setBillingProfileId(createProjectRequest.projectName, profileModel.getId)
    }
```

I think the for comprehension is slightly more readable, but the desugared approach is fine with me as well if you feel strongly about it.

Reply:
I don't feel strongly about it if there's a reason… I just couldn't figure out what the reason was. I am now reading through https://docs.scala-lang.org/overviews/core/futures.html.